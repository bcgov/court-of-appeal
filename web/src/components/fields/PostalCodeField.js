import React, { Component } from 'react';
import {INVALID_POSTAL_CODE_MSG} from '../../helpers/constants';

class PostalCodeField extends Component {

    constructor(props) {
        super(props);

        this.state = {
            className: "form-field",
            title: '',
            errorMessage: ''
        };
        this.handleFieldChange = this.handleFieldChange.bind(this);
        this.validate = this.validate.bind(this);
        this.clearClass = this.clearClass.bind(this);
    }

    render() {
        return (
            <div>
                <input
                id={this.props.id}
                value={this.props.value || ''}
                onChange={this.handleFieldChange}
                maxLength={7}
                size={30}
                title={this.state.title}
                className={this.state.className}
                onBlur={this.validate}
                onFocus={this.clearClass}
                placeholder="Valid Canadian postal code"
                readOnly={this.props.readOnly}
                name={this.props.name}
                 />
                <div className="error-message">{this.state.errorMessage}</div>
            </div>
        );
    }

    componentDidMount() {
        if (this.props.value) {
            let isValid = this.validate();
            if (isValid && this.props.value) {
                this.setState({className: "form-field preloaded"});
            }
        }
    }

    handleFieldChange(e) {
        if (!this.props.readOnly) {
            let value = e.target.value;
            value = value.toUpperCase().replace(/[^\dA-Z ]/g,"").replace(/[ ]+/g," ");
            if(value.length === 3 && value.length > this.props.value.length) {
                value = value + (" ");
            }
            e.target.value =  value;
            this.props.handleFieldChange(e);
        }
    }

    /**
     * If, when the field loses focus, it's not empty or a proper postal code, give the field an error color, and give
     * an error title message so they can see what's up when they hover.
     */
    validate() {
        if (!this.props.value || this.props.value.match(/^[a-zA-Z]\d[a-zA-Z] \d[A-Z]\d$/) != null) {
            this.setState({className: "form-field", title: ""}, this.props.validate(true, 'postalCode'));
            return true;
        } else {
            this.setState({
                className: "form-field invalid-field",
                title: INVALID_POSTAL_CODE_MSG,
                errorMessage: INVALID_POSTAL_CODE_MSG
            }, this.props.validate(false, 'postalCode'));
            return false;
        }
    }

    clearClass() {
        this.setState({className: "form-field", title: "", errorMessage: ""});
    }
}
export default PostalCodeField;
