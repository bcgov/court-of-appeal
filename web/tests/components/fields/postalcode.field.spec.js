import React from 'react';
import PostalCodeField from '../../../src/components/fields/PostalCodeField';
import renderer from 'react-test-renderer';

test('should only allow valid Canadian postal code characters', ()=> {

    let value = '';
    let handleFieldChange = (e) => { value = e.target.value; };

    const postalCode = renderer.create(
        <PostalCodeField id="number-field" value={value}
                    handleFieldChange={handleFieldChange.bind(this)}/>,
    );
    let tree = postalCode.toJSON();
    expect(tree).toMatchSnapshot();

    let e = { target: { value: "42"}};

    let pc = postalCode.getInstance();

    pc.handleFieldChange(e);
    expect(value).toMatch('42');

    e = { target: { value: "bah" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('BAH');

    e = { target: { value: "heyBabaReeBear 123" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('HEYBABA');

    e = { target: { value: "V0N 1J22" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('V0N 1J2');

    e = { target: { value: "V0N   2" } };
    pc.handleFieldChange(e);
    expect(value).toMatch('V0N 2');

});
