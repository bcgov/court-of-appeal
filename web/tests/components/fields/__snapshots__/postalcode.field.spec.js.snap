// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should only allow valid Canadian postal code characters 1`] = `
<div>
  <input
    className="form-field"
    id="number-field"
    maxLength={7}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Valid Canadian postal code"
    size={30}
    title=""
    value=""
  />
  <div
    className="error-message"
  >
    
  </div>
</div>
`;
